### PR TITLE
Allow upgrading between different BCrypt encodings

### PR DESCRIPTION
**Pull request by larsgrefer**
This will allow the upgradeEncoding mechanism to upgrade from a lower-strength bcrypt to a higher-strength bcrypt encoding.

### CODE REVIEW

File: crypto/src/main/java/org/springframework/security/crypto/password/DelegatingPasswordEncoder.java [hunk 1]
```diff
@@ -217,9 +217,15 @@ private String extractId(String prefixEncodedPassword) {
 	}
 
 	@Override
-	public boolean upgradeEncoding(String encodedPassword) {
-		String id = extractId(encodedPassword);
-		return !this.idForEncode.equalsIgnoreCase(id);
+	public boolean upgradeEncoding(String prefixEncodedPassword) {
```
Comment (rwinch) [inline crypto/src/main/java/org/springframework/security/crypto/password/DelegatingPasswordEncoder.java]: Please add a test for these changes

File: crypto/src/main/java/org/springframework/security/crypto/bcrypt/BCryptPasswordEncoder.java [hunk 2]
```diff
@@ -125,6 +122,24 @@ public boolean matches(CharSequence rawPassword, String encodedPassword) {
 		return BCrypt.checkpw(rawPassword.toString(), encodedPassword);
 	}
 
+	@Override
+	public boolean upgradeEncoding(String encodedPassword) {
+		if (encodedPassword == null || encodedPassword.length() == 0) {
+			logger.warn("Empty encoded password");
```
Comment (rwinch) [inline crypto/src/main/java/org/springframework/security/crypto/bcrypt/BCryptPasswordEncoder.java]: Please add additional tests

**Comment by larsgrefer** [comment]
I'm not sure what the default behavior for invalid inputs should be.

`org.springframework.security.crypto.password.DelegatingPasswordEncoder#upgradeEncoding` returns `true` for all invalid inputs (null, no parseable id, unknown id) while `org.springframework.security.crypto.password.PasswordEncoder#upgradeEncoding` always returns false.

The default of `false` makes sense, because it prevents useless re-encodings on every login.

How should `org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder#upgradeEncoding` behave for invalid inputs (null or non-bcrypt)? `true`, `false` or `IllegalArgumentException` ?

**Comment by rwinch** [comment]
Good question @larsgrefer
 
`DelegatingPasswordEncoder` returns true for null, no id, unknown id because we expect its configuration to change over time. All we care about is that the value is encoded with the configured PasswordEncoder. Note that these are actually valid as there is the potential to fallback if there is an unknown id, null id, etc.

`BCryptPasswordEncoder` should return false for a null input because it will blow up if it is encoded. For non-bcrypt it should fail with an Exception.

**Comment by rwinch** [comment]
Thanks for the updates @larsgrefer. It appears that [the tests](https://api.travis-ci.org/v3/job/551839898/log.txt) fail.

```
org.springframework.security.crypto.bcrypt.BCryptPasswordEncoderTests > upgradeFromNullOrEmpty FAILED
    org.junit.ComparisonFailure at BCryptPasswordEncoderTests.java:191
```

**Comment by larsgrefer** [comment]
@rwinch The tests are fixed now

**Comment by rwinch** [comment]
Thanks for all your work on this PR @larsgrefer! This is now merged into master via d3d6a8743e9b17653320d967cc42d2e4dc50adba I added a little polish to the tests via 742df2cd1dc30a929674e341a1366effb1327364

**Pull request closed without merging**